### PR TITLE
improve local user password docs

### LINKED ISSUE #3516: API: Field label and doc issues with setting password in `local_idp_user_create` and `local_idp_user_set_password` requests
Currently the password value in the request payload is presented as follows in the API docs:
image
The top-level comment "password used to log in" should perhaps be changed to: "set or invalidate user password".

The `user_password_value` field is actually the operation: set password vs invalidate password, not the password value.

The `details` field is where the password value should be set. Apparently this and the field above had their labels swapped during implementation (?).

Finally, it's unclear if the "Passwords may be subject to additional constraints" remark is helpful at all because we don't provide a way to set/apply password policies.

### PR DESCRIPTION
**Pull request by davepacheco**
Fixes #3516.

### CODE REVIEW

**Comment by davepacheco** [comment]
@askfongjojo what do you think of this version?  It looks like this:

Screenshot 2023-07-07 at 11 12 55 AM
Screenshot 2023-07-07 at 11 12 59 AM

File: openapi/nexus.json [hunk 1]
```diff
@@ -12768,34 +12768,34 @@
...
           },
           {
             "description": "Invalidates any current password (disabling password authentication)",
             "type": "object",
             "properties": {
-              "user_password_value": {
+              "mode": {
                 "type": "string",
                 "enum": [
                   "invalid_password"
```
Comment (zephraph) [inline openapi/nexus.json:12793]: Perhaps this should be `invalidate_password` instead? `invalid_password` implies a password that's not valid, not something that will remove passwords.

**Comment by zephraph** [comment]
so the input is

```json
{
  "mode": "password",
  "value": "adsfasdl;fajsdf;lajds"
}
```

or 

```json
{
  "mode": "invalid_password"
}
```

---

I would almost prefer an input of

```json
{
  "password": "asdflak;jsdf;"
}
```

or 

```json
{
  "invalidate_password": true
}
```

**Comment by davepacheco** [comment]
> so the input is
> 
> ```json
> {
>   "mode": "password",
>   "value": "adsfasdl;fajsdf;lajds"
> }
> ```
> 
> or
> 
> ```json
> {
>   "mode": "invalid_password"
> }
> ```
> 
> I would almost prefer an input of
> 
> ```json
> {
>   "password": "asdflak;jsdf;"
> }
> ```
> 
> or
> 
> ```json
> {
>   "invalidate_password": true
> }
> ```

I think this is not allowed by openapi-lint.  I initially tried removing the annotation and got:

```
---- integration_tests::commands::test_nexus_openapi stdout ----
writing temp config: /dangerzone/omicron_tmp/test_commands_config.1968.0
thread 'integration_tests::commands::test_nexus_openapi' panicked at 'Problem with type UserCreate: Mismatched types between subschemas; this is often due to enums with different data payloads and can be resolved using serde adjacent tagging.
this schema's type
Object(
    ObjectType {
        properties: {
            "password": Reference {
                reference: "#/components/schemas/Password",
            },
        },
        required: [
            "password",
        ],
        additional_properties: Some(
            Any(
                false,
            ),
        ),
        min_properties: None,
        max_properties: None,
    },
)
differs from this
String(
    StringType {
        format: Empty,
        pattern: None,
        enumeration: [
            Some(
                "invalid_password",
            ),
        ],
        min_length: None,
        max_length: None,
    },
)

For more info, see https://github.com/oxidecomputer/openapi-lint#type-mismatch

Problem with type UserPassword: Mismatched types between subschemas; this is often due to enums with different data payloads and can be resolved using serde adjacent tagging.
this schema's type
Object(
    ObjectType {
        properties: {
            "password": Reference {
                reference: "#/components/schemas/Password",
            },
        },
        required: [
            "password",
        ],
        additional_properties: Some(
            Any(
                false,
            ),
        ),
        min_properties: None,
        max_properties: None,
    },
)
differs from this
String(
    StringType {
        format: Empty,
        pattern: None,
        enumeration: [
            Some(
                "invalid_password",
            ),
        ],
        min_length: None,
        max_length: None,
    },
)

For more info, see https://github.com/oxidecomputer/openapi-lint#type-mismatch', nexus/tests/integration_tests/commands.rs:123:5
```

File: nexus/types/src/external_api/params.rs [hunk 2]
```diff
@@ -276,7 +276,7 @@ pub struct SiloCreate {
 pub struct UserCreate {
     /// username used to log in
     pub external_id: UserId,
-    /// password used to log in
+    /// describes how to set the user's login password
```
Comment (ahl) [inline nexus/types/src/external_api/params.rs:279]:
```suggestion
    /// how to set the user's login password
```

maybe? in the spirit of
- less is more
- 80 columns in the CLI help menu

File: nexus/types/src/external_api/params.rs [hunk 3]
```diff
@@ -276,7 +276,7 @@ pub struct SiloCreate {
 pub struct UserCreate {
```
Comment (ahl) [inline nexus/types/src/external_api/params.rs:276]: What do we want the JSON to look like? Today it's something like this:
```json
{
  "external_id": "xyz",
  "password": {
    "mode": "password",
    "value": "l1admin"
  }
}

{
  "external_id": "xyz",
  "password": {
    "mode": "invalid_password"
  }
}
```

That seems good. Is this any better?

```json
{
  "external_id": "xyz",
  "password": "l1admin"
}

{
  "external_id": "xyz",
}
```

The absence of `"password"` or a value of `null` would indicate that the password should be unset? It's the best alternative I can think of, and I hate it.

Or maybe:
```json
{
  "external_id": "xyz",
  "mode": "password",
  "value": "l1admin"
}

{
  "external_id": "xyz",
  "mode": "invalid_password"
}
```

... but that means we'd have a Rust type like this:

```rust
#[serde(tag = "mode")]
pub enum UserCreate {
    Password {
        pub external_id: UserId,
        pub value: String,
    }
    InvalidPassword {
        pub external_id: UserId,
    }
}
```

Enumeration options to say: I like what you've chosen and can't think of something better

File: openapi/nexus.json (hunk 1 above)
Comment (ahl) [reply openapi/nexus.json:12793]: Or maybe even "no_login" or "login_disallowed"?

**Review by ahl — APPROVED** [review]
thanks for pointing me at this

**Comment by ahl** [comment]
> > I would almost prefer an input of
> > ```json
> > {
> >   "password": "asdflak;jsdf;"
> > }
> > ```
> > or
> > ```json
> > {
> >   "invalidate_password": true
> > }
> > ```
> 
> I think this is not allowed by openapi-lint. I initially tried removing the annotation and got:

Removing the `#[serde(...)]`? Yeah, that would make it an "externally" tagged enum which openapi-lint squawks about when you mix a variant with a type and a simple variant. However, I think @zephraph is actually saying something like:

```rust
enum UserPassword {
    Password(Password),
    InvalidatePassword(AlwaysTrue)`
}
/// Serializes as just the value true 
struct AlwaysTrue;
// needs Serialize / Deserialize impls
impl JsonSchema for AlwaysTrue {
    fn json_schema(..) -> Schema {
        serde_json::from_value::<Schema>(json! {
            "enum": [ true ]
        })
    }
    fn is_referenceable() -> bool {
       false
    }
}
```

**Pull request merged**